ext/include/opentelemetry/ext/zpages/threadsafe_span_data.h

```diff
@@ -0,0 +1,173 @@
+#include <mutex>
+#include <unordered_map>
+#include <vector>
+
+#include "opentelemetry/core/timestamp.h"
+#include "opentelemetry/nostd/string_view.h"
+#include "opentelemetry/sdk/trace/recordable.h"
+#include "opentelemetry/trace/canonical_code.h"
+#include "opentelemetry/trace/span.h"
+#include "opentelemetry/sdk/trace/span_data.h"
+#include "opentelemetry/trace/span_id.h"
+#include "opentelemetry/trace/trace_id.h"
+
+using opentelemetry::sdk::trace::AttributeConverter;
+using opentelemetry::sdk::trace::SpanDataAttributeValue;
+namespace trace_api = opentelemetry::trace;
+
+OPENTELEMETRY_BEGIN_NAMESPACE
+namespace ext {
+namespace zpages {
+
+/**
+ * This class is a threadsafe version of span data used for zpages in OT
+ */
+class ThreadsafeSpanData final : public opentelemetry::sdk::trace::Recordable {
+ public:
+  /**
+   * Get the trace id for this span
+   * @return the trace id for this span
+   */
+  opentelemetry::trace::TraceId GetTraceId() const noexcept {
+    std::lock_guard<std::mutex> lock(mutex_);
+    return trace_id_;
+  }
+
+  /**
+   * Get the span id for this span
+   * @return the span id for this span
+   */
+  opentelemetry::trace::SpanId GetSpanId() const noexcept {
+    std::lock_guard<std::mutex> lock(mutex_);
+    return span_id_;
+  }
+
+  /**
+   * Get the parent span id for this span
+   * @return the span id for this span's parent
+   */
+  opentelemetry::trace::SpanId GetParentSpanId() const noexcept {
+    std::lock_guard<std::mutex> lock(mutex_);
+    return parent_span_id_;
+  }
+
+  /**
+   * Get the name for this span
+   * @return the name for this span
+   */
+  opentelemetry::nostd::string_view GetName() const noexcept {
+    std::lock_guard<std::mutex> lock(mutex_);
+    return name_;
+  }
+
+  /**
+   * Get the status for this span
+   * @return the status for this span
+   */
+  opentelemetry::trace::CanonicalCode GetStatus() const noexcept {
+    std::lock_guard<std::mutex> lock(mutex_);
+    return status_code_;
+  }
+
+  /**
+   * Get the status description for this span
+   * @return the description of the the status of this span
+   */
+  opentelemetry::nostd::string_view GetDescription() const noexcept {
+    std::lock_guard<std::mutex> lock(mutex_);
+    return status_desc_;
+  }
+
+  /**
+   * Get the start time for this span
+   * @return the start time for this span
+   */
+  opentelemetry::core::SystemTimestamp GetStartTime() const noexcept {
+    std::lock_guard<std::mutex> lock(mutex_);
+    return start_time_;
+  }
+
+  /**
+   * Get the duration for this span
+   * @return the duration for this span
+   */
+  std::chrono::nanoseconds GetDuration() const noexcept {
+    std::lock_guard<std::mutex> lock(mutex_);
+    return duration_;
+  }
+
+  /**
+   * Get the attributes for this span
+   * @return the attributes for this span
+   */
+  const std::unordered_map<std::string, SpanDataAttributeValue> &GetAttributes()
+      const noexcept {
+    std::lock_guard<std::mutex> lock(mutex_);
```

Comment:
So this is not really protecting anything - it's a false sense of safety. You return a reference to map, yeah - you get this from under the lock. But the code that got a reference - it is not under the lock, so depending on use-case - this is racy...

Reply:
@maxgolov is right. Here you could simply implement your own thread-safe getter that returns a copy of the attribute map. The `Recordable` interface doesn't specify any getters.

You could probably also optimize this further for the zPages use case. I guess you don't need all getters before the span is ended. You would only need to protect those getters and setters that are used by zPages before the span is passed to the `OnEnd` processor method.

Reply:
I think all our needed because we copy data like the ids and duration etc and convert them to strings which can easily be converted to a json object for the front end to render.
I missed the & I'll make the change thanks!